// express-openapi-validator/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ono_1 = require("ono");
const _ = require("lodash");
const middlewares = require("./middlewares");
const openapi_context_1 = require("./framework/openapi.context");
const openapi_spec_loader_1 = require("./framework/openapi.spec.loader");
class OpenApiValidator {
    constructor(options) {
        this.validateOptions(options);
        if (options.unknownFormats == null)
            options.unknownFormats === true;
        if (options.coerceTypes == null)
            options.coerceTypes = true;
        if (options.validateRequests == null)
            options.validateRequests = true;
        if (options.validateResponses == null)
            options.validateResponses = false;
        if (options.validateResponses === true) {
            options.validateResponses = {
                removeAdditional: false,
            };
        }
        if (options.validateRequests === true) {
            options.validateRequests = {
                allowUnknownQueryParameters: false,
            };
        }
        this.options = options;
        const spec = new openapi_spec_loader_1.OpenApiSpecLoader({
            apiDoc: this.options.apiSpec,
        }).load();
        this.context = new openapi_context_1.OpenApiContext(spec, options.ignorePaths);
    }
    install(app) {
        this.installPathParams(app);
        this.installMetadataMiddleware(app);
        this.installMultipartMiddleware(app);
        const components = this.context.apiDoc.components;
        if (components && components.securitySchemes) {
            this.installSecurityMiddleware(app);
        }
        if (this.options.validateRequests) {
            this.installRequestValidationMiddleware(app);
        }
        if (this.options.validateResponses) {
            this.installResponseValidationMiddleware(app);
        }
    }
    installPathParams(app) {
        const pathParams = [];
        for (const route of this.context.routes) {
            if (route.pathParams.length > 0) {
                pathParams.push(...route.pathParams);
            }
        }
        // install param on routes with paths
        for (const p of _.uniq(pathParams)) {
            app.param(p, (req, res, next, value, name) => {
                const { pathParams } = req.openapi;
                if (pathParams) {
                    // override path params
                    req.params[name] = pathParams[name] || req.params[name];
                }
                next();
            });
        }
    }
    installMetadataMiddleware(app) {
        app.use(middlewares.applyOpenApiMetadata(this.context));
    }
    installMultipartMiddleware(app) {
        app.use(middlewares.multipart(this.context, this.options.multerOpts));
    }
    installSecurityMiddleware(app) {
        const securityMiddleware = middlewares.security(this.context, this.options.securityHandlers);
        app.use(securityMiddleware);
    }
    installRequestValidationMiddleware(app) {
        const { coerceTypes, unknownFormats, validateRequests } = this.options;
        const { allowUnknownQueryParameters } = (validateRequests);
        const requestValidator = new middlewares.RequestValidator(this.context.apiDoc, {
            nullable: true,
            coerceTypes,
            removeAdditional: false,
            useDefaults: true,
            unknownFormats,
            allowUnknownQueryParameters,
        });
        const requestValidationHandler = (req, res, next) => requestValidator.validate(req, res, next);
        app.use(requestValidationHandler);
    }
    installResponseValidationMiddleware(app) {
        const { coerceTypes, unknownFormats, validateResponses } = this.options;
        const { removeAdditional } = validateResponses;
        const responseValidator = new middlewares.ResponseValidator(this.context.apiDoc, {
            nullable: true,
            coerceTypes,
            removeAdditional,
            unknownFormats,
        });
        app.use(responseValidator.validate());
    }
    validateOptions(options) {
        if (!options.apiSpec)
            throw ono_1.default('apiSpec required');
        const securityHandlers = options.securityHandlers;
        if (securityHandlers != null) {
            if (typeof securityHandlers !== 'object' ||
                Array.isArray(securityHandlers)) {
                throw ono_1.default('securityHandlers must be an object or undefined');
            }
        }
        const unknownFormats = options.unknownFormats;
        if (typeof unknownFormats === 'boolean') {
            if (!unknownFormats) {
                throw ono_1.default("unknownFormats must contain an array of unknownFormats, 'ignore' or true");
            }
        }
        else if (typeof unknownFormats === 'string' &&
            unknownFormats !== 'ignore' &&
            !Array.isArray(unknownFormats))
            throw ono_1.default("unknownFormats must contain an array of unknownFormats, 'ignore' or true");
    }
}
exports.OpenApiValidator = OpenApiValidator;
